numpy/_core/src/multiarray/item_selection.c

```diff
@@ -1064,10 +1062,29 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
              * so the input array is not changed
              * before the error is called
              */
-            flags |= NPY_ARRAY_ENSURECOPY;
+            is_newarr = 1;
+        }
+
+        if (!PyArray_EquivTypes(dtype, PyArray_DESCR(out))) {
+            is_newarr = 1;
         }
+
         Py_INCREF(dtype);
-        obj = (PyArrayObject *)PyArray_FromArray(out, dtype, flags);
+        if (is_newarr) {
+            obj = (PyArrayObject *)PyArray_NewFromDescr(Py_TYPE(out),
+                                                        dtype,
+                                                        multi->nd,
+                                                        multi->dimensions,
+                                                        NULL, NULL, 0,
+                                                        (PyObject *)out);
+        }
+        else {
+            int flags = NPY_ARRAY_CARRAY |
+                        NPY_ARRAY_WRITEBACKIFCOPY |
+                        NPY_ARRAY_FORCECAST;
+
+            obj = (PyArrayObject *)PyArray_FromArray(out, dtype, flags);
```

Comment:
This branch should now never copy, so the only thing it should be doing is check for `PyArray_FailUnlessWriteable`.  That also means you don't need the writeback-if-copy branch below.

(We can check that earlier also, since it doesn't make sense to wait until the last copy for this to fail in the other branch either.)

Reply:
Makes sense, thanks! Addressed this.